tool/generator/translator.dart

```diff
@@ -384,80 +423,136 @@ class _PartialInterfacelike {
+          if (!_shouldGenerateMember(fieldName)) break;
+          properties.add(_Field(_MemberName(fieldName), field.idlType,
+              mdnInterface?.propertyFor(fieldName),
+              isRequired: field.required));
           break;
         case 'maplike':
         case 'setlike':
         case 'iterable':
-          members.add(member);
+          // TODO(srujzs): Generate members for these types.
           break;
         default:
           throw Exception('Unrecognized member type $type');
       }
     }
   }
 
+  /// Given the [declaredInheritance] by the IDL, find the closest supertype
+  /// that is actually generated, and set the inheritance equal to that type.
+  void _setInheritance(String? declaredInheritance) {
+    if (declaredInheritance == null) return;
+    final translator = Translator.instance!;
+    while (declaredInheritance != null) {
+      if (translator.markTypeAsUsed(declaredInheritance)) {
+        inheritance = declaredInheritance;
+        break;
+      } else {
+        declaredInheritance = (translator
+                ._typeToDeclaration[declaredInheritance] as idl.Interfacelike)
+            .inheritance;
+      }
+    }
+  }
+
+  /// Given a [memberName] and whether it [isStatic], return whether it is a
+  /// member that should be emitted according to the compat data.
+  bool _shouldGenerateMember(String memberName, {bool isStatic = false}) {
+    // Compat data only exists for interfaces and namespaces. Mixins and
+    // dictionaries should always generate their members.
+    if (type != 'interface' && type != 'namespace') return true;
+    final interfaceBcd =
+        Translator.instance!.browserCompatData.retrieveInterfaceFor(name)!;
+    final bcd = interfaceBcd.retrievePropertyFor(memberName,
+        // Compat data treats namespace members as static, but the IDL does not.
+        isStatic: isStatic || type == 'namespace');
+    final shouldGenerate = bcd?.shouldGenerate;
+    if (shouldGenerate != null) return shouldGenerate;
+    // Events can bubble up to the window, document, or other elements. In the
+    // case where we have no compatibility data, we assume that an event can
+    // bubble up to this interface and support the event handler.
+    if (!isStatic && BrowserCompatData.isEventHandlerSupported(memberName)) {
+      return true;
+    }
+    // TODO(srujzs): Sometimes compatibility data can be up or down the type
+    // hierarchy, so it may be worth checking supertypes and subtypes. In
+    // practice, it doesn't seem to make a difference in the output.
+    return false;
+  }
+
   void update(idl.Interfacelike interfacelike) {
     assert((name == interfacelike.name && type == interfacelike.type) ||
         interfacelike.type == 'interface mixin');
     assert(interfacelike.inheritance == null || inheritance == null,
         'An interface should only be defined once.');
-    inheritance ??= interfacelike.inheritance;
+    _setInheritance(interfacelike.inheritance);
```

Comment:
Is this done to skip-over interfaces that are excluded, even though a leaf and a grand-parent interface are emitted? Does that happen a lot in practice?

I wonder if it would be less breaking to emit the intermediate interfaces, but empty, rather than changing the type-hierarchy if one of the intermediate classes ends up being emitted in a future version.

Reply:
Yes, exactly. And it only occurs for two superinterfaces that are not emitted: `Accelerometer` and `WebGLObject`, each of which affect 2 and 13 subinterfaces, respectively.

What breakages are you thinking might occur if we add the intermediate type later? I would assume that would be non-breaking whereas potentially removing the intermediate type would be breaking (although I doubt `WebGLObject` and `Accelerometer` are going anywhere).

Reply:
A scenario that comes to mind is if we have non-experimental APIs that use the skipped-over type in its signature. For example, if we have: `external void doFoo(WebGLObject arg)`, we'd be changing it to `external void doFoo(JSObject arg)`? Then if `WebGLObject` is reintroduced, such API would have a breaking change

That said, given your observation yesterday that breaking changes are unavoidable anyways, I have no problem keeping what you have here.

Reply:
Ah I see, yeah that would be a breaking change. I think either decision would make this a breaking change, so for now, I'll avoid emitting the empty extension type.